include/deal.II/base/hdf5.h

```diff
@@ -0,0 +1,712 @@
+ * pcout << "Local no collective cause :"
+ *       << dataset.local_no_collective_cause<std::string>() << std::endl;
+ * pcout << "Global no collective cause :"
+ *       << dataset.global_no_collective_cause<std::string>() << std::endl;
+ * #endif
+ * @endcode
+ *
+ * @author Daniel Garcia-Sanchez, 2018
+ */
+namespace HDF5
+
+{
+  /**
+   * Base class for the HDF5 objects.
+   *
+   * @author Daniel Garcia-Sanchez, 2018
+   */
+  class HDF5Object
+  {
+  protected:
+    HDF5Object(const std::string name, bool mpi);
+
+  public:
+    enum class Mode
+    {
+      create,
+      open
+    };
+
+    /**
+     * Reads an attribute. T can be float, double, std::complex<float>,
+     * std::complex<double>, int, unsigned int, bool or std::string. Note that
+     * the encoding of std::string is UTF8 in order to be compatible with
+     * python3.
+     *
+     * Datatype conversion takes place at the time of a read or write and is
+     * automatic. See the <a
+     * href="https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDatatypes%2FHDF5_Datatypes.htm%23TOC_6_10_Data_Transferbc-26&rhtocid=6.5_2">Data
+     * Transfer: Datatype Conversion and Selection</a>  section in the HDF5
+     * User's Guide.
+     */
+    template <typename T>
+    T
+    attr(const std::string attr_name);
+
+    /**
+     * Writes an attribute. T can be float, double, std::complex<float>,
+     * std::complex<double>, int, unsigned int, bool or std::string. Note that
+     * the encoding of std::string is UTF8 in order to be compatible with
+     * python3.
+     *
+     * Datatype conversion takes place at the time of a read or write and is
+     * automatic. See the <a
+     * href="https://support.hdfgroup.org/HDF5/doc/UG/HDF5_Users_Guide-Responsive%20HTML5/index.html#t=HDF5_Users_Guide%2FDatatypes%2FHDF5_Datatypes.htm%23TOC_6_10_Data_Transferbc-26&rhtocid=6.5_2">Data
+     * Transfer: Datatype Conversion and Selection</a>  section in the HDF5
+     * User's Guide.
+     */
+    template <typename T>
+    void
+    write_attr(const std::string attr_name, const T value);
+
+
+    const std::string name;
+
+  protected:
+    /**
+     * HDF5 identifier for the objects File, Group and DataSet
+     */
+    std::shared_ptr<hid_t> hdf5_reference;
+
+    /**
+     * If true use parallel HDF5, if false use serial HDF5.
+     */
+    const bool mpi;
+
+    /**
+     * Returned value of the last HDF5 C library call.
+     *
+     * The HDF5 calls return a non-negative value if successful; otherwise they
+     * return a negative value.
+     *
+     * Calls such as
+     * [H5DWrite](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5D.html#Dataset-Write)
+     * use this variable to store the returned value. Other calls such as
+     * [H5Pcreate](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-Create)
+     * don't use this variable to store the returned value.
+     */
+    herr_t ret;
```

Comment:
That is good point. Originally I defined `herr_t` as local. But that creates lot of warnings in release mode. Let's take the example:
``` c++
ret = H5Aread(attr, H5T_NATIVE_INT, &int_value);
Assert(ret >= 0, ExcMessage("Error at H5Aread"));
````
If `ret` is local then the compiler will complain because the variable is not used in release mode. I see the following options. Let me know what you prefer and if you have a better option:
- Make ret a member of the class (current option)
- Make ret local and use `AssertThrow` instead of `Assert` everywhere. I don't know if there will be a performance penalty. My guess is that the performance penalty will be very small.
- Use C++17 maybe_unused. Not possible because deal.ii uses C++11

Reply:
Just like @masterleinad says, write this as
```
  ret = H5Aread (...);
  (void)ret;
  Assert (ret>0, ...);
```
This avoids the warning and still allows you to make the variable local.